refactor(devtools): filtering providers by token within "injector tree" tab

ducin opened the pull request:
In some injectors, there are LOTS of providers, making it slightly inconvenient to search for a certain one. This commit introduces a search-by-token for providers of a specific injector.

## PR Checklist

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit

## PR Type

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [X] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

Some injectors within the application might include **_lots_** of providers, making it inconvenient to search for them effectively, like below:

<img width="822" alt="Zrzut ekranu 2023-12-1 o 18 33 55" src="https://github.com/angular/angular/assets/375027/ad7c881d-f044-4ff2-9809-b9d2b6ea8878">

(scroll is so long)

<img width="817" alt="Zrzut ekranu 2023-12-1 o 18 35 07" src="https://github.com/angular/angular/assets/375027/5df19285-0096-443a-910f-5922047dc93d">

## What is the new behavior?

This PR introduces a simple search-by-token, allowing to filter effectively. Clicking the 'close' icon clears the input.

<img width="626" alt="image" src="https://github.com/angular/angular/assets/375027/9ab53aa8-bd86-40b1-a134-fbfee12403d8">

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

Commit:
refactor(devtools): filtering providers by token within "injector tree" tab

In some injectors, there are LOTS of providers, making it slightly inconvenient to search for a certain one. This commit introduces a search-by-token for providers of a specific injector.

ducin commented:
PS the original idea came from my angular training participant who commented on the "Injector Tree" tab:
> _well, cool, but it would be difficult to scroll through the whole list of providers..._
so here it is!

AleksanderBodurri commented:
Love it! Will review this later today. Thanks @ducin 🙏

JeanMeche commented:
Padding/margin of the title doesn't look great, can you improve it ? :)

JeanMeche commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts:
```diff
@@ -20,7 +22,16 @@ import {Events, MessageBus, SerializedInjector, SerializedProviderRecord} from '
+            <mat-form-field appearance="fill">
+              <mat-label>Search by token</mat-label>
+              <input matInput
+                placeholder="provider token"
```
nit:
```suggestion
                placeholder="Provider token"
```

ducin replied:
fixed

JeanMeche commented:
Per @eneajaho's suggestion on X, let's a filter for the type also !

Commit:
fixup! refactor(devtools): filtering providers by token within "injector tree" tab

changes:
- placeholder uppercased @JeanMeche
- second filter - by type

eneajaho commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts:
```diff
@@ -108,20 +134,40 @@ import {Events, MessageBus, SerializedInjector, SerializedProviderRecord} from '
   standalone: true,
-  imports: [NgIf, NgForOf, MatTableModule, MatIconModule, MatTooltipModule],
+  imports: [
+    NgIf, NgForOf, MatTableModule, MatIconModule, MatTooltipModule, MatInputModule, MatSelectModule,
```
Do we need NgIf and NgFor here?

ducin replied:
@eneajaho  the code was there before, but:
```
  <div *ngIf="injector" class="injector-providers">
```
so I guess NgIf - yes. Not sure if `@for` requires NgForOf... does it?

JeanMeche replied:
The new control flow is a compiler featuer. It doesn't require the directives anymore !

ducin replied:
done - both if and for removed, ngIf -> @if

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts:
```diff
@@ -18,9 +21,32 @@ import {Events, MessageBus, SerializedInjector, SerializedProviderRecord} from '
   template: `
         <h1>
           Providers for {{ injector?.name  }}
+          {{ searchToken() }} {{ searchType() }}
```
I think we can remove the titles here. The input field values are good indicators already of what we've selected

ducin replied:
oops... done!

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts:
```diff
@@ -18,9 +21,32 @@ import {Events, MessageBus, SerializedInjector, SerializedProviderRecord} from '
         <div *ngIf="injector" class="injector-providers">
-            <table *ngIf="providers.length > 0" mat-table [dataSource]="providers" class="mat-elevation-z4">
+            <mat-form-field appearance="fill">
+              <mat-label>Search by token</mat-label>
```
Lets space out these input fields so that they are not so close together

<img width="823" alt="Screenshot 2023-12-01 at 8 41 47 PM" src="https://github.com/angular/angular/assets/39253660/fd396268-392c-41ae-9064-39d7cf36a391">

ducin replied:
done via css class

Commit:
fixup! refactor(devtools): filtering providers by token within "injector tree" tab

Review fixes:
- removed unnecessary directive imports
- spaces between fields
- removed useless header parts

Commit:
fixup! refactor(devtools): filtering providers by token within "injector tree" tab

fix: removed accidentally left console.log statement

AleksanderBodurri reviewed (approved):
thank you @ducin 🙏  great work!

dylhunn commented:
This PR was merged into the repository by commit ba8dc2843a49b69a034881c38c8d85135c0bf3d6.
